Abins: fix q-dependence of estimated high-order terms

ajjackson opened the pull request:
**Description of work.**

High-order phonon contributions are estimated by repeated convolution of the fundamental spectrum against the highest calculated order. (As this may be the first order, we refer to the method as 'autoconvolution'. But strictly speaking only the first step is actually "auto"...)

In Abins energy and q are related by the instrument/angle geometry, so the q^2 dependence of the spectrum is "baked in" to the calculation over energies at each angle. However, as high order spectra should scale with (q^2n) (where q is the quantum order), it is not correct to rely on the q information in the fundamental spectrum. Instead, the approach should be:

- Compute the spectra at a fixed q (1 recip. angstrom is convenient...)

- Apply the q^2n terms after q-independent line shapes are determined by convolution

Here we do this slightly inefficiently to fit the existing Abins toolchain (further refactors to be considered...) and

- Compute the order-1 spectrum as a function of q (with and without the order-1 Debye-Waller factor) and, if desired, the order-2 spectrum.

- Remove the q-dependence from the non-DW-corrected spectra by division

- Use the fixed-q order-1 spectrum as the basis for auto-convolution

- Apply (q^2n / n!) intensity correction to each kernel

- Apply q-dependent Debye-Waller factors exp(-tr(A) q^2/3) to all orders

- Replace order-1 data with the previously-computed spectrum (including a mode-dependent Debye-Waller term.)

This replaces an implementation in which the n! correction was neglected and convolution kernels were normalised to 1. It is believed that some other implementations of this method perform convolution
with normalised kernels; however it seems more convenient to leave the (tr(B)/3) term in the convolution than to remove it and later apply (tr(B)/3)^n. Perhaps this comes at a cost in precision?

### Work-in-progress checklist
- [x] Update sdata autoconvolution test, consider removing normalised mode
- [x] Check scaling is consistent when multiple k-points are included
- [x] Check results broadly agree with other implementations
- [ ] ~~Sort out a bug: Abins complains when calculating order-1 (and convolving the rest) if order-2 data already exists~~ **Opened issue #33598**

**To test:**

<!-- Instructions for testing. -->
Abins outputs that _don't_ involve convolution are quite well tested already.
Results with convolution have changed; currently there is no test for results with convolution. A wider range of tests should be added for the Abins powder calculations - but they would not prove much about this PR so I suggest it is a separate piece of work.
Until then, manual testing will show the improvement: run Abins with and without the Autoconvolution option ticked. Both runs should work, and the former should have a nice spectrum of overtones, accumulating into a broadly flat shelf. Examples are shown in the Issue discussion below.

This will change user results (for the better!) and requires a release note. The release note has been written, by may need moving/adapting to the new system?

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

ajjackson committed:
Abins: add multiplication methods to SData

This will allow energy-dependent corrections (which may actually be
q-dependent...) to be applied to calculated S intensities with clean
syntax.

The core logic of spowdersemiempiricalcalculator needs to juggle a few
different scenarios so it is helpful if the data manipulation can be
clear and terse - making it easier to focus on the logic.

ajjackson committed:
Abins: apply appropriate q-dependence to autoconvolution spectra

High-order phonon contributions are estimated by repeated convolution
of the fundamental spectrum against the highest calculated order. (As
this may be the first order, we refer to the method as
'autoconvolution'. But strictly-speaking only the first step is
actually "auto"...)

In Abins energy and q are related by the instrument/angle geometry,
so the q^2 dependence of the spectrum is "baked in" to the calculation
over energies at each angle. However, as high order spectra should
scale with (q^2n) (where q is the quantum order), it is not correct
to rely on the q information in the fundamental spectrum. Instead,
the approach should be:

- Compute the spectra at a fixed q (1 recip. angstrom is
  convenient...)

- Apply the q^2n terms after q-independent line shapes are determined
  by convolution

Here we do this slightly inefficiently to fit the existing Abins
toolchain (further refactors to be considered...) and

- Compute the order-1 spectrum as a function of q (with and without
  the order-1 Debye-Waller factor) and, if desired, the order-2 spectrum.

- Remove the q-dependence from the non-DW-corrected spectra by
  division

- Use the fixed-q order-1 spectrum as the basis for auto-convolution

- Apply (q^2n / n!) intensity correction to each kernel

- Apply q-dependent Debye-Waller factors exp(-tr(A) q^2/3) to all
  orders

- Replace order-1 data with the previously-computed
  spectrum (including a mode-dependent Debye-Waller term.)

This replaces an implementation in which the n! correction was
neglected and convolution kernels were normalised to 1. It is believed
that some other implementations of this method perform convolution
with normalised kernels; however it seems more convenient to leave
the (tr(B)/3) term in the convolution than to remove it and later
apply (tr(B)/3)^n. Perhaps this comes at a cost in precision?

ajjackson commented:
The results look fairly reasonable at this stage:

- At high energy a broad roughly-flat shelf forms, which is expected from experiment and from other implementations
- Estimating order-2 by convolution gives broadly similar results to calculating it by a more exact method, and the same general intensity.

![benzene-o1-conv_improved](https://user-images.githubusercontent.com/2511383/143475838-65581d30-bca6-4e6f-af2d-fe2d1ccb0586.png

![benzene-o2-conv_improved](https://user-images.githubusercontent.com/2511383/143475844-fbbc7e89-9b42-49c6-bc39-a6ee3ea4a967.png)
)

ajjackson commented:
Compare this with the current *main* branch implementation. We can see that without the (q^2n) correction the high-order terms are squashed flat by the (exponentially-decaying) Debye-Waller factor.

![benzene-o2-conv_broken](https://user-images.githubusercontent.com/2511383/143476752-0f2d61ba-8988-4f35-bc9a-a2f194f950fe.png)

ajjackson committed:
Abins: Further q-dependence refactoring

- Calculate at fixed-q directly (rather than calculate with
  q-dependence then remove it.)
- Remove code entirely for (disabled) orders 3, 4. (It seems to be
  mathematically equivalent to self-convolution, except for the
  prefactors which lack a clear origin.)
- Rework order-1 data flow a bit:
  - Calculate with DW after higher orders are deal with, just before
    inserting into SData
  - Move DW calculation outside intensity calculation for more
    consistency with order 2

ajjackson commented:
I have refactored this a little more to properly separate the calculation of fundamentals for convolution and for the results (including mode-dependent DW). This seems to have had the side-effect of cleaning up the convolution kernel and we now have a smoother result that agrees well with other implementations.

![benzene-o2-conv_improved-2](https://user-images.githubusercontent.com/2511383/143911080-e9725d3d-553a-40e0-8484-cf3353b20fb6.png)

Compared to an O-Climax simulation of the same data with equivalent settings:
![benzene-oclimax](https://user-images.githubusercontent.com/2511383/143911200-07f9afc0-0427-4a0f-a6fc-e47a03d12d6a.png)

The qualitative agreement is now quite acceptable with a consistent smooth flat shelf. There is an overall scale factor difference, but we probably shouldn't worry about that until temperature terms and k-point weighting have been re-examined.

The most striking difference at this point is a difference in the convention regarding which scattering angle is "forward" and which is "backward". I suspect this is my mistake when implementing the Abins settings; it needs to be checked with the instrument scientists. But that is a separate bug/issue to be resolved.

ajjackson commented:
Here are the spectra from Castep calculations with different numbers of input q-points. The scaling appears to be handled consistently. I'd like to take a closer look at how that works at some point, but this provides some assurance for now.

![abins-scaling-q](https://user-images.githubusercontent.com/2511383/144632072-d6ab6229-48af-4907-a754-90f96b81c62b.png)

(The system is CuH as this has a nice small unit cell for DFPT calculations with different numbers of samples.)

ajjackson commented:
If we focus on a high quantum order, though, the picture becomes less convincing.

![CuH-111_H_quantum_event_6-1](https://user-images.githubusercontent.com/2511383/144633994-5d552eca-09f9-429a-afbe-7b141edeb849.png)

ajjackson committed:
Abins: release notes update

ajjackson commented:
I have explored the q-dependence a bit more carefully and find that it is possible to create discrepancies by adding duplicate k-points to data *if they are not weighted correctly*. If the phonon k-point weights sum to 1, things seem to stay robust and reasonable.

I'm not sure this is the convention for all data formats; the safest option would be to enforce it when creating an Abins KpointsData object.

ajjackson committed:
Abins: normalise k-point weights

When k-points sum to more than 1 this messes up the maths for
self-convolution. There seem to be some discrepancies at order 2 as
well. The safest option seems to be normalising at import time; then
from that point the rest of Abins knows what to expect.

ajjackson commented:
We can see that this breaks the CRYSTAL test because the CRYSTAL sample data uses integer q-point weights (i.e. summing to > 1).

ajjackson commented:
![MgO-CRYSTAL_Mg_quantum_event_1](https://user-images.githubusercontent.com/2511383/152203777-5d85e52c-0ada-413c-827d-1430bad3b99f.png)
![MgO-CRYSTAL_Mg_quantum_event_2](https://user-images.githubusercontent.com/2511383/152203794-d93c7fdd-e541-479b-a87f-98877019b66e.png)

Comparing branches, the normalisation change does not seem to affect orders 1 & 2 of the actual calculated MgO spectrum. So while the KpointsData weights need updating, user results shouldn't be affected. (Except for the corrected order 3+ spectrum...)

ajjackson committed:
Abins: update MgO CRYSTAL test reference (change phonon kpt weights)

Only the kpt weights have been changed, this was achieved by editing
the existing file rather than generating a new one.

ajjackson committed:
Abins: update mgo-GX_LoadCRYSTAL test data

k-points weighting scheme has changed: weights are normalised at
input, so change in the AbinsData objects compared by unit tests.

File was diffed against previous version and changes are as expected:
- New k-point weights
- Advanced parameters changed. (Expected, these are ignored in test.)

The atomic_displacements files have changed format but data should be
the same, so these are not updated. The test should pass against the
old ones...

ajjackson committed:
Abins: update CRYSTAL MgO-GX test displacement data

It turns out that the Gamma-point displacements do need updating to
match the other updated test file (mgo-GX_LoadCRYSTAL_data.txt). This
is due to an existing change in Abins; zero/negative-frequency modes
are no longer omitted from the loaded AbinsData, and instead are
filtered out during processing. This makes data shapes more regular
and possible loss of scientific data to a somewhat arbitrary
thresholding process.

To avoid completely regenerating test data (and hiding unintended
changes that might have occurred at the same time) the test was
written to be somewhat robust against old reference files, screening
out low frequencies with the _cull_imaginary_modes
staticmethod. However, this method was written such that the reference
frequencies and displacements were still expected to be /consistent/.

As in this case we have regenerated the *_data.txt file, to capture a
genuine change in how k-points are handled, the consistency was broken
at the Gamma-point, where there are (as expected) three imaginary
modes. So we need to update the displacements file at this k-point to
match.

I have checked the numbers by hand, and the relationship between the
files is as expected: new rows are added to the start of the file
which represent rigid lattice translations, and the existing values
are unchanged. (The linebreaks are also lost, but this should be
inconsequential.)

ajjackson committed:
Correct MgO-222 test data weights

In 179d7517df I manually updated this test file to ensure that only
the expected data was changed: the k-point weights.

Inexplicably, I changed them to 1/6 instead of 1/8, despite there
clearly being 8 k-points in the system, which are - according to
CRYSTAL - equally weighted.

Hopefully this will rectify that.

ajjackson commented:
Fixing the CRYSTAL tests took longer than expected, due to a mixture of a relatively subtle problem and a really silly one :sweat_smile: .

ajjackson committed:
Abins: remove unused autoconvolution scaling/normalisation

Now that the maths has been corrected, we don't need to normalise the
autoconvolution kernels numerically so can cut out some logic.

ajjackson committed:
Migrate a release note to new format

AnthonyLim23 commented on docs/source/release/v6.4.0/Indirect/Algorithms/Bugfixes/33064.rst:
```diff
@@ -0,0 +1,4 @@
+The Abins algorithm has reworked to account for q-dependence
+correctly when forming 3+ phonon events by self-convolution. This
+changes the results obtained when using the "Autoconvolution"
+option, which now results in the expected high-frequency "wing".
```
would it be worth including a plot of before and after the change to highlight the new wing?

ajjackson replied:
Good idea! Is there any guidance on how that should be done in the new system? Would a file in this directory and a relative link in the RST work?

AnthonyLim23 replied:
@sf1919 is this in your docs?

ajjackson replied:
Ok, done! Not sure how to preview it, I guess problems if any will come up at release time?

sf1919 replied:
Leave it as is. If there are any problems it will be picked up once we have the compilation process in place. If changes do need to be made I will make them and update documentation accordingly.

AnthonyLim23 commented on scripts/abins/spowdersemiempiricalcalculator.py:
```diff
@@ -576,10 +606,36 @@ def _calculate_s_over_threshold(s=None, freq=None, coeff=None):
+        Calculate mode-dependent Debye-Waller factor for the first order quantum event for one atom.
+        :param q2: squared values of momentum transfer vectors
+        :param frequencies: frequencies for which transitions occur
+        :param indices: (unused) array which stores information how transitions can be decomposed in terms of fundamentals
```
will this be used in the future?

ajjackson replied:
Probably not, it can come out 😄 

(It's an artefact of when this function would be passed around and needed a consistent interface with higher orders.)

AnthonyLim23 commented on scripts/abins/spowdersemiempiricalcalculator.py:
```diff
@@ -593,21 +649,7 @@ def _calculate_order_one(self, *, q2: np.ndarray, frequencies: np.ndarray,
-            return s, s_with_dw
-        else:
-            return s
+        return s
```
could just do return on line 651

ajjackson commented on scripts/abins/spowdersemiempiricalcalculator.py:
```diff
@@ -576,10 +606,36 @@ def _calculate_s_over_threshold(s=None, freq=None, coeff=None):
+        else:
+            coth = 1.0 / np.tanh(frequencies * CM1_2_HARTREE
+                                 / (2.0 * self._temperature * K_2_HARTREE))
+        dw = np.exp(-q2 * (a_trace + 2.0 * trace_ba / b_trace) / 5.0 * coth * coth)
```
This is fine in Python, `if` blocks do not create a new scope.

ajjackson replied:
```python
#! usr/bin/env python3                                                         |
                                                                               |
if True:                                                                       |
    x = "It's ok"                                                              |
                                                                               |
print(x)                                                                       |

AnthonyLim23 replied:
its a force of habit

ajjackson replied:
This C program is fine:
```c
#include <stdio.h>

int main() {
  if ( 1 ) {
    char text[] = "It's ok\n";
    printf("%s", text);    
  }

  return 0;
}
﻿```
but this one raises a compiler error:
```c
#include <stdio.h>

int main() {
  if ( 1 ) {
    char text[] = "It's not ok\n";
  }

  printf("%s", text);    

  return 0;
}
```

ajjackson committed:
Reviewer comments: clean up

Thanks @AnthonyLim23

- indices is vestigal and can come out
- return without needless assignment

ajjackson committed:
Reviewer comments: changelog wording, illustration

Thanks @AnthonyLim23, @sf1919

I don't see a way to preview if this image link is correct for the new
release not infrastructure. I guess somebody will check it eventually?

AnthonyLim23 reviewed (dismissed):
Code looks good and the autoconvultion appears to work as expected

DavidFair commented on scripts/abins/kpointsdata.py:
```diff
@@ -57,6 +64,14 @@ def __init__(self, *, frequencies: np.ndarray, atomic_displacements: np.ndarray,
             raise ValueError("Invalid value of weights.")
+
+        if not isclose(np.sum(weights), 1.0):
+            if logger is None:
+                logger = mantid_logger
```
Would it be better to do the DI unconditionally, rather than when the sum of weights is not approx 1.0?
Long-term I can see confusion / debugging sessions arising when developers have to have specific weights for their logging to work in other places

ajjackson replied:
Good point! And it's generally more conventional to put this sort of "argument handling" close to the top of a function/method block, so that is where one would go looking for it.

DavidFair reviewed (commented):
Just one point, otherwise looks good. Well commented and nice readable code :+1:

ajjackson committed:
Abins KpointsData: load logger unconditionally

Thanks @DavidFair for review suggestion. This removes potential
confusion during debugging/testing as logger is set depending on
values of input data.
